File: src/pushNotification/push.js
const express = require("express");
const bodyParser = require("body-parser");
const Notification = require("./index");
const router = require("express").Router();
const { prepareConnection } = require("../../database/conection");

router.use(bodyParser.urlencoded({ extended: false }));

router.use(bodyParser.json());

router.post("/notification/token", function (req, res) {
  //res.send("Sending Notification to One user....")

  //kpi.listarLastMonht(req,res)
  prepareConnection(0)
    .then((resolve) => {
      const registrationToken = req.body.registrationToken;
      console.log(registrationToken);
      //const message = req.body.message;

      let respuesta = {
        fecha: resolve[0]["hist_date"].toISOString(),
        kpiEstimado: resolve[0]["hist_kpiEstimado"],
        kpiReal: resolve[0]["hist_kpiReal"],
      };

      const notification_options = {
        priority: "high",
        timeToLive: 60 * 60 * 24,
      };

      const message_notification = {
        notification: {
          title: " \u{1F602} ESTADISTICAS KPI BBVA ",
          body: `Fecha: ${respuesta.fecha.slice(
            0,
            respuesta.fecha.indexOf("T")
          )} 
              Kpi estimado: ${respuesta.kpiEstimado} %
              Kpi real: ${Number(respuesta.kpiReal).toFixed(4)} %
            `,
        },
      };

      resServerMessage(
        registrationToken,
        message_notification,
        notification_options
      );

      res.status(200).json({
        status: 200,
        message: "Ok",
        response: {
          Status: "Sending Notification to One user....",
        },
      });

      console.log(resolve);
    })
    .catch((reject) => {
      res.status(500).json({
        status: 500,
        message: "Internal Server Error",
        response: reject,
      });
      console.log(reject);
    });
});

router.post("/notification/topic", function (req, res) {
  //const message = req.body.message;

  prepareConnection(0)
    .then((resolve) => {
      const registrationTopic = req.body.registrationTopic;
      console.log(registrationTopic);
      //const message = req.body.message;

      let respuesta = {
        fecha: resolve[0]["hist_date"].toISOString(),
        kpiEstimado: resolve[0]["hist_kpiEstimado"],
        kpiReal: resolve[0]["hist_kpiReal"],
        emoji : ""
      };

      const notification_options = {
        priority: "high",
        timeToLive: 60 * 60 * 24,
      };
      

      if(Number(respuesta.kpiReal).toFixed(4) > respuesta.kpiEstimado){
        respuesta.emoji = `\u{1F44F}`
        console.log("Esta con manitos arriba")
      }else if(Number(respuesta.kpiReal).toFixed(4) == respuesta.kpiEstimado){
        respuesta.emoji = `\u{1F603}`
        console.log("Esta feliz") 
      }else if(Number(respuesta.kpiReal).toFixed(4) < respuesta.kpiEstimado){
        console.log("Esta triste") 
        respuesta.emoji = `\u{1F61F}` 
      }

      const message_notification = {
        notification: {
          title:  ` ${respuesta.emoji} ESTADISTICAS KPI BBVA`,
          body: `Fecha: ${respuesta.fecha.slice(
            0,
            respuesta.fecha.indexOf("T")
          )} 
                Kpi estimado: ${Number((respuesta.kpiEstimado)*100).toFixed(2)}%
                Kpi real: ${((Number(respuesta.kpiReal).toFixed(4))*100).toFixed(2)}%
              `,
        },
      };

      resServerMessageTopic(
        registrationTopic,
        message_notification,
        notification_options
      );

      res.status(200).json({
        status: 200,
        message: "Ok",
        response: {
          Status: "Sending Notification to One user....",
        },
      });

      console.log(resolve);
    })
    .catch((reject) => {
      res.status(500).json({
        status: 500,
        message: "Internal Server Error",
        response: reject,
      });
      console.log(reject);
    });
});

async function resServerMessage(tokenOrTopic, message_notification, options) {
  const respuesta = await Notification.sendMessageToToken(
    tokenOrTopic,
    message_notification,
    options
  );
  console.log(respuesta.results[0]);
  console.log(respuesta);
}

async function resServerMessageTopic(
  tokenOrTopic,
  message_notification,
  options
) {
  const respuesta = await Notification.sendMessageToTopic(
    tokenOrTopic,
    message_notification,
    options
  );
  //console.log(respuesta.results[0])
  console.log(respuesta);
}

module.exports = router;

File: src/pushNotification/push.py
from flask import Flask, request, jsonify
from firebase_admin import messaging
from database.conection import prepareConnection
import os
from datetime import datetime

app = Flask(__name__)
app.config['SECRET_KEY'] = os.environ.get('API_ACCES_KEY')

def generate_notification_options():
    return messaging.AndroidConfig(
        priority='high',
        ttl=datetime.utcnow() + timedelta(days=1)
    )

def generate_message_notification(title, body):
    return messaging.Notification(
        title=title,
        body=body
    )

def send_message_to_token(registration_token, notification, options):
    message = messaging.Message(
        notification=notification,
        token=registration_token,
        android=options
    )
    return messaging.send(message)

def send_message_to_topic(topic, notification, options):
    message = messaging.Message(
        notification=notification,
        topic=topic,
        android=options
    )
    return messaging.send(message)

@app.route('/notification/token', methods=['POST'])
def send_notification_to_token():
    try:
        registration_token = request.json.get('registrationToken', '')
        if not registration_token:
            return jsonify({'status': 400, 'message': 'Bad Request'}), 400

        resolve = prepare_connection(0)
        notification_options = generate_notification_options()
        emoji = ''
        kpi_real = float(resolve[0]["hist_kpiReal"])
        kpi_estimado = float(resolve[0]["hist_kpiEstimado"])

        if kpi_real > kpi_estimado:
            emoji = '\U0001F44F'
            print("Esta con manitos arriba")
        elif kpi_real == kpi_estimado:
            emoji = '\U0001F603'
            print("Esta feliz")
        elif kpi_real < kpi_estimado:
            print("Esta triste")
            emoji = '\U0001F61F'

        notification_title = f'{emoji} ESTADISTICAS KPI BBVA'
        notification_body = f'Fecha: {resolve[0]["hist_date"].strftime("%Y-%m-%d")} \nKpi estimado: {kpi_estimado * 100:.2f}%\nKpi real: {kpi_real * 100:.2f}%'
        notification = generate_message_notification(notification_title, notification_body)

        send_message_to_token(registration_token, notification, notification_options)

        return jsonify({'status': 200, 'message': 'Ok', 'response': {'Status': 'Sending Notification to One user....'}})
    except Exception as e:
        return jsonify({'status': 500, 'message': 'Internal Server Error', 'response': str(e)}), 500

@app.route('/notification/topic', methods=['POST'])
def send_notification_to_topic():
    try:
        registration_topic = request.json.get('registrationTopic', '')
        if not registration_topic:
            return jsonify({'status': 400, 'message': 'Bad Request'}), 400

        resolve = prepare_connection(0)
        notification_options = generate_notification_options()
        emoji = ''
        kpi_real = float(resolve[0]["hist_kpiReal"])
        kpi_estimado = float(resolve[0]["hist_kpiEstimado"])

        if kpi_real > kpi_estimado:
            emoji = '\U0001F44F'
            print("Esta con manitos arriba")
        elif kpi_real == kpi_estimado:
            emoji = '\U0001F603'
            print("Esta feliz")
        elif kpi_real < kpi_estimado:
            print("Esta triste")
            emoji = '\U0001F61F'

        notification_title = f'{emoji} ESTADISTICAS KPI BBVA'
        notification_body = f'Fecha: {resolve[0]["hist_date"].strftime("%Y-%m-%d")} \nKpi estimado: {kpi_estimado * 100:.2f}%\nKpi real: {kpi_real * 100:.2f}%'
        notification = generate_message_notification(notification_title, notification_body)

        send_message_to_topic(registration_topic, notification, notification_options)

        return jsonify({'status': 200, 'message': 'Ok', 'response': {'Status': 'Sending Notification to One user....'}})
    except Exception as e:
        return jsonify({'status': 500, 'message': 'Internal Server Error', 'response': str(e)}), 500

if __name__ == '__main__':
    app.run(port=8080, debug=True)
